metals/src/main/scala/scala/meta/metals/MetalsLanguageServer.scala

```diff
@@ -0,0 +1,272 @@
+import scala.meta.internal.metals.UserConfiguration
+import scala.meta.internal.metals.clients.language.MetalsLanguageClient
+import scala.meta.internal.metals.clients.language.NoopLanguageClient
+import scala.meta.internal.metals.logging.LanguageClientLogger
+import scala.meta.internal.metals.logging.MetalsLogger
+import scala.meta.internal.metals.{MetalsLanguageServer => MetalsLspService}
+import scala.meta.io.AbsolutePath
+import scala.meta.metals.ServerState.ShuttingDown
+import scala.meta.metals.lsp.DelegatingScalaService
+import scala.meta.metals.lsp.LanguageServer
+import scala.meta.metals.lsp.ScalaLspService
+
+import org.eclipse.lsp4j._
+
+/**
+ * According to the spec, the server waits for the `initialize` request to be
+ * sent. After that, the server sends the `initialized` notification to the
+ * client. Next, the server is fully working and can receive requests and
+ * notifications from the client.
+ */
+sealed trait ServerState
+object ServerState {
+  case object Started extends ServerState
+  final case class Initialized(service: MetalsLspService) extends ServerState
+  final case class ShuttingDown(service: MetalsLspService) extends ServerState
+}
+
+/**
+ * Scala Language Server implementation.
+ *
+ * @param ec
+ *   execution context for futures.
+ * @param sh
+ *   scheduled executor service for scheduling tasks.
+ */
+class MetalsLanguageServer(
+    ec: ExecutionContextExecutorService,
+    buffers: Buffers = Buffers(),
+    redirectSystemOut: Boolean = true,
+    charset: Charset = StandardCharsets.UTF_8,
+    time: Time = Time.system,
+    initialConfig: MetalsServerConfig = MetalsServerConfig.default,
+    initialUserConfig: UserConfiguration = UserConfiguration.default,
+    progressTicks: ProgressTicks = ProgressTicks.braille,
+    bspGlobalDirectories: List[AbsolutePath] =
+      BspServers.globalInstallDirectories,
+    sh: ScheduledExecutorService = Executors.newSingleThreadScheduledExecutor(),
+    isReliableFileWatcher: Boolean = true,
+    mtagsResolver: MtagsResolver = MtagsResolver.default(),
+    onStartCompilation: () => Unit = () => (),
+    classpathSearchIndexer: ClasspathSearch.Indexer =
+      ClasspathSearch.Indexer.default,
+) extends LanguageServer {
+
+  ThreadPools.discardRejectedRunnables("MetalsLanguageServer.sh", sh)
+  ThreadPools.discardRejectedRunnables("MetalsLanguageServer.ec", ec)
+
+  private implicit val executionContext: ExecutionContextExecutorService = ec
+
+  // AtomicReference's set and get have the same semantics as volatile variables.
+  // because we only use get and set, there is no need to use AtomicReference.
+  @volatile
+  private var serverState: ServerState = ServerState.Started
+
+  @volatile
+  private var languageClient: MetalsLanguageClient = NoopLanguageClient
+
+  private val cancelables = new MutableCancelable()
+  private val isCancelled = new AtomicBoolean(false)
+  private val isLanguageClientConnected = new AtomicBoolean(false)
```

Comment:
Again, excuse the naive question, but related to your comment above. In this situation for example we only use `set` and `get`, but are still using an `AtomicBoolean` over a `@volatile`. How come?

Reply:
good point, I'm not being consistent. As I wrote earlier, it's probably easier to switch to `Atomic*` and just forget about `@volatile`